### PR TITLE
refactoring(glimmer): Extract utils from printer

### PR DESCRIPTION
**Pull request by dcyriller**
#### Description

Extract a few glimmer printer utils into a file, similar to what is done for others languages.

#### Checklist
- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-handlebars/utils.js [hunk 1]
```diff
@@ -11,6 +11,87 @@ function isGlimmerComponent(node) {
...
+function isWhitespaceNode(node) {
+  return node.type === "TextNode" && !/\S/.test(node.chars);
+}
+
+function isNodeOfSomeType(node, types) {
+  if (node) {
+    return types.some(type => node.type === type);
+  }
+  return false;
+}
```
Comment (fisker) [inline src/language-handlebars/utils.js]:
```suggestion
function isNodeOfSomeType(node, types) {
  return node && types.some(type => node.type === type);
}
```

Comment (dcyriller) [reply src/language-handlebars/utils.js]: I applied your suggestion

File: src/language-handlebars/utils.js [hunk 2]
```diff
@@ -11,6 +11,84 @@ function isGlimmerComponent(node) {
...
+
+function isPrettierIgnoreNode(node) {
+  if (!isNodeOfSomeType(node, ["MustacheCommentStatement"])) {
+    return false;
+  }
+
+  return (
+    typeof node.value === "string" && node.value.trim() === "prettier-ignore"
+  );
+}
```
Comment (fisker) [inline src/language-handlebars/utils.js]:
```suggestion
function isPrettierIgnoreNode(node) {
  return (
    isNodeOfSomeType(node, ["MustacheCommentStatement"]) &&
    typeof node.value === "string" &&
    node.value.trim() === "prettier-ignore"
  );
}
```

File: src/language-handlebars/utils.js [hunk 3]
```diff
@@ -11,6 +11,84 @@ function isGlimmerComponent(node) {
...
+
+function getNextNode(path) {
+  const node = path.getValue();
+  const parentNode = path.getParentNode(0);
+
+  const children = parentNode.children || parentNode.body;
+  if (children) {
+    const nodeIndex = children.indexOf(node);
+    if (nodeIndex < children.length) {
```
Comment (fisker) [inline src/language-handlebars/utils.js]: This seems can't be false?

Comment (fisker) [reply src/language-handlebars/utils.js]: Is this mean to check `< .length - 1`?

Comment (fisker) [reply src/language-handlebars/utils.js]: but still unnessary `children[nodeIndex + 1]` also get `undefined`

File: src/language-handlebars/utils.js [hunk 4]
```diff
@@ -11,6 +11,84 @@ function isGlimmerComponent(node) {
...
+      const nextNode = children[nodeIndex + 1];
+      return nextNode;
+    }
+  }
+}
```
Comment (fisker) [inline src/language-handlebars/utils.js]:
```suggestion
function getNextNode(path) {
  const node = path.getValue();
  const parentNode = path.getParentNode(0);
  const children = parentNode.children || parentNode.body;
  return children[children.indexOf(node) + 1];
}
```

Comment (fisker) [reply src/language-handlebars/utils.js]: We are not check `index === -1`, I don't think need check `children` either

Comment (dcyriller) [reply src/language-handlebars/utils.js]: My initial goal was more to extract these utils than to refactor them. But there is definitely work to do in this area. 😄

Comment (dcyriller) [reply src/language-handlebars/utils.js]: Nice one

Comment (fisker) [reply src/language-handlebars/utils.js]: Should be safe

Comment (fisker) [reply src/language-handlebars/utils.js]: `getPreviousNode` checked `nodeIndex > 0` , I think we should also check it in `getPreviousNode`

Comment (fisker) [reply src/language-handlebars/utils.js]: I updated the suggestion

File: src/language-handlebars/utils.js [hunk 5]
```diff
@@ -11,6 +11,84 @@ function isGlimmerComponent(node) {
...
+
+  const children = parentNode && (parentNode.children || parentNode.body);
+  if (children) {
+    const nodeIndex = children.indexOf(node);
+    if (nodeIndex > 0) {
+      const previousNode = children[nodeIndex - lookBack];
+      return previousNode;
+    }
+  }
+}
```
Comment (fisker) [inline src/language-handlebars/utils.js]:
```suggestion
function getPreviousNode(path, lookBack = 1) {
  const node = path.getValue();
  const parentNode = path.getParentNode(0) || {};
  const children = parentNode.children || parentNode.body || [];
  const index = children.indexOf(node)
  return index !== -1 && children[index - lookBack];
}
```

File: src/language-handlebars/utils.js (hunk 4 above)
Comment (fisker) [reply src/language-handlebars/utils.js]: How about ...

```js
const getNextNode = node => getPreviousNode(node, -1);
```

Comment (dcyriller) [reply src/language-handlebars/utils.js]: > getPreviousNode checked nodeIndex > 0 , I think we should also check it in getNextNode, that one also checked parentNode

agreed

File: src/language-handlebars/utils.js (hunk 5 above)
Comment (dcyriller) [reply src/language-handlebars/utils.js]: I have applied your suggestions. It unlocks a few good improvements. Do you mind having a look at my last commit?

File: src/language-handlebars/utils.js [hunk 6]
```diff
@@ -1,16 +1,80 @@
...
+
+function getNextNode(path) {
+  const node = path.getValue();
+  const parentNode = path.getParentNode(0);
+  const children = parentNode.children || parentNode.body || [];
+  const index = children.indexOf(node);
+  return index !== -1 && children[index + 1];
+}
+
```
Comment (fisker) [inline src/language-handlebars/utils.js]: Looks like my review get lost, some how,

Make a `getSiblingNode(path, offset)`, and `getNextNode`, `getPreviousNode` as shortcuts

Comment (dcyriller) [reply src/language-handlebars/utils.js]: CI was red because of a typo

Good advice, I'll add it.

Comment (dcyriller) [reply src/language-handlebars/utils.js]: Done, neat refactoring in the end. Thanks for your help :)

File: src/language-handlebars/utils.js [hunk 7]
```diff
@@ -1,16 +1,80 @@
...
-  }
+  return (
+    isNodeOfSomeType(node, ["ElementNode"]) &&
+    typeof node.tag === "string" &&
+    (isUppercase(node.tag[0]) || node.tag.includes("."))
+  );
+}
+
+function isWhitespaceNode(node) {
+  return isNodeOfSomeType(node, ["TextNode"]) && !/\S/.test(node.chars);
```
Comment (fisker) [inline src/language-handlebars/utils.js]:
```suggestion
  return isNodeOfSomeType(node, ["TextNode"]) && /\s/.test(node.chars);
```

Comment (fisker) [reply src/language-handlebars/utils.js]: Seems don't need negative, right?

Comment (fisker) [reply src/language-handlebars/utils.js]: Wait, not equal, if all `\s`

Comment (fisker) [reply src/language-handlebars/utils.js]: Sorry, I think my suggestion is not right.

Comment (fisker) [reply src/language-handlebars/utils.js]:
```js
!/\S/.test('')
true
/\s/.test('')
false
```

**Comment by fisker** [comment]
@dcyriller need undo last one, my fault.

File: src/language-handlebars/utils.js (hunk 7 above)
Comment (fisker) [reply src/language-handlebars/utils.js]: /^\s*$/ should be equal, but old one better

Comment (dcyriller) [reply src/language-handlebars/utils.js]: Removed the last commit 😉

**Pull request merged**